kentikapi/internal/synthetics/test.go

```diff
@@ -0,0 +1,652 @@
+	}
+
+	if t.Id == "" {
+		return nil, fmt.Errorf("empty test ID in response payload")
+	}
+
+	settings, err := testSettingsFromPayload(t)
+	if err != nil {
+		return nil, fmt.Errorf("convert response payload to test settings: %w", err)
+	}
+
+	createdBy := userInfoPtrFromPayload(t.CreatedBy)
+	if createdBy == nil {
+		return nil, errors.New("createdBy field is nil in response payload")
+	}
+
+	return &synthetics.Test{
+		Name:          t.Name,
+		Status:        synthetics.TestStatus(t.Status.String()),
+		Settings:      settings,
+		ID:            t.Id,
+		Type:          synthetics.TestType(t.Type),
+		CreateDate:    t.Cdate.AsTime(),
+		UpdateDate:    t.Edate.AsTime(),
+		CreatedBy:     *createdBy,
+		LastUpdatedBy: userInfoPtrFromPayload(t.LastUpdatedBy),
+	}, nil
+}
+
+func testSettingsFromPayload(t *syntheticspb.Test) (synthetics.TestSettings, error) {
+	if t.Settings == nil {
+		return synthetics.TestSettings{}, fmt.Errorf("empty test settings")
+	}
+	ts := t.Settings
+
+	definition, err := testDefinitionFromPayload(t)
+	if err != nil {
+		return synthetics.TestSettings{}, err
+	}
+
+	hs, err := healthSettingsFromPayload(ts.HealthSettings)
+	if err != nil {
+		return synthetics.TestSettings{}, err
+	}
+
+	return synthetics.TestSettings{
+		Definition:           definition,
+		AgentIDs:             ts.AgentIds,
+		Tasks:                taskTypesFromPayload(ts.Tasks),
+		Health:               hs,
+		Ping:                 pingSettingsFromPayload(ts.Ping),
+		Traceroute:           tracerouteSettingsFromPayload(ts.Trace),
+		Period:               time.Duration(ts.Period) * time.Second,
+		Family:               synthetics.IPFamily(ts.Family.String()),
+		NotificationChannels: ts.NotificationChannels,
+	}, nil
+}
+
+// nolint: gocyclo
+func testDefinitionFromPayload(t *syntheticspb.Test) (synthetics.TestDefinition, error) {
+	switch synthetics.TestType(t.Type) {
+	case synthetics.TestTypeIP:
+		return ipTestDefinitionFromPayload(t.Settings)
+	case synthetics.TestTypeNetworkGrid:
+		return networkGridTestDefinitionFromPayload(t.Settings)
+	case synthetics.TestTypeHostname:
+		return hostnameTestDefinitionFromPayload(t.Settings)
+	case synthetics.TestTypeAgent:
+		return agentTestDefinitionFromPayload(t.Settings)
+	case synthetics.TestTypeNetworkMesh:
+		return networkMeshTestDefinitionFromPayload(t.Settings)
+	case synthetics.TestTypeFlow:
+		return flowTestDefinitionFromPayload(t.Settings)
+	case synthetics.TestTypeURL:
+		return urlTestDefinitionFromPayload(t.Settings)
+	case synthetics.TestTypePageLoad:
+		return pageLoadTestDefinitionFromPayload(t.Settings)
+	case synthetics.TestTypeDNS:
+		return dnsTestDefinitionFromPayload(t.Settings)
+	case synthetics.TestTypeDNSGrid:
+		return dnsGridTestDefinitionFromPayload(t.Settings)
+	default:
+		return nil, fmt.Errorf("invalid test type: %v", t.Type)
+	}
+}
+
+func ipTestDefinitionFromPayload(ts *syntheticspb.TestSettings) (synthetics.TestDefinition, error) {
```

Comment:
Can go support general (introspection based) [construction of SDK objects based on protobuf data](https://github.com/kentik/community_sdk_python/blob/1bac5553fc04ba3d536602a372cd380c3ba32f87/kentik_api_library/kentik_api/synthetics/synth_tests/base.py#L56) like Python does?

Reply:
Yes, I think it is possible, using the `reflect` package. There are several caveats, however:
1. I renamed some fields to be more user-friendly. That could be reverted or we could use struct tags on public models to specify PB field name.
2. Some PB fields are translated to user-friendly types, e.g. `string` to `ip.Addr`, `string` with number to `int`, `string` with duration to `time.Duration`. That could be reverted or we could use struct tags on public models to specify custom translation function.
3. I merged `activation.time_unit` and `activation.time_window` PB fields to single `Activation.TimeWindow` field of type time.Duration. That could be reverted or we could find some other mechanism to keep that state.
4. Possibly other aspects
5. There is a library for copying fields from one struct to other, but it might not be flexible enough: https://github.com/jinzhu/copier/  Its custom converters look messy: https://github.com/jinzhu/copier/pull/115

It would be quite hard to write and read. It might be more maintainable in the long run, but I am skeptical. I definitely don't want to do that in this PR. Do you want me to create a task for that?

Reply:
Thanks for the explanation. We do not need to do it if the benefits are comparatively small.